File restructuring and renaming

KAClough opened the pull request:
This PR aims to clear the following issues:

1. Having both CCZ4Vars and CCZ4Variables is confusing #105 - changed the latter to CCZ4StateVariables. With other proposed changes the former should not be needed anyway, but adding "State" felt better anyway.

2. General restructuring of the folders per issues:
- Remove InitialConditions folder #102
- Partly addresses: Re-organise directory structure #6 and Random updates to GRChombo naming/formats #92

I have added the new folders `MathsTools` `IOTools` and `GridTools` and tried to split the various miscellaneous tools between them. Open to suggestions as to whether this could be done another way. 

I moved a lot of junk into the External folder in the base folder. Again I am open to renaming this but could not think of anything more useful than "Misc".

3. Renames PositiveChiAndAlpha -> PositiveChiAndLapse #99

4. This partly fixes Replace bad tagging criteria with better ones #104 by deleting the old ones. We can add the better ones with the porting of specific examples.

KAClough committed:
Starts restructing of files, probably breaks stuff

KAClough committed:
Change CCZ4Variables name and clang format

KAClough committed:
runs clang format v19

KAClough committed:
Change directory structure

KAClough committed:
Fixes the lint thing

KAClough committed:
Fixes naming of positive chi and lapse class

KAClough committed:
Fixes tests that I broke

the-florist commented:
Hi all, 
The PR looks quite good overall, I like the three categories you've chosen (Math, IO and Grid tools) and I think it's really clear that the Kerr example and the BH initial conditions are all in the same folder. It also leaves more room in the Examples if we want to add new matter examples in the future! I've just got some minor suggestions for sorting the files:

1. AMReXParameters.hpp got placed in GRTeclynCore but as it's a file dealing with text input, it could also go into IOTools.
2. ArrayTools also got put in GRTeclynCore but could go into GridTools, as for example the Tensor class definition is there too and Array/Tensor functionality seems like it could be related.
3. Perhaps NullBCFill could also go into GridTools as the other Boundary Condition files are there.

Of course these are just suggestions, feel free to take or leave them :)

mirenradia commented on Source/BlackHoles/BinaryBH.hpp:
Can you rename this class and file name to `BinaryBHInitialData`?

KAClough replied:
Done

mirenradia commented on Source/BlackHoles/BoostedBH.hpp:
Can you rename this class and file name to `BoostedBHInitialData`?

KAClough replied:
Done

mirenradia commented on Source/BlackHoles/KerrBH.hpp:
Can you rename this class and file name to `KerrBHInitialData`?

KAClough replied:
Done

mirenradia commented on Source/BlackHoles/Make.package:
Whilst we're editing this file, can you make it 1 file per line and in
alphabetical order (as is the case for most/some of the other `Make.package` files)?

KAClough replied:
Done

mirenradia commented on Source/GRTeclynCore/AlwaysInline.hpp:
We should get rid of this file and just use `AMREX_FORCE_INLINE` instead. Let's
defer that to another PR though.

KAClough replied:
Made a new issue.

mirenradia commented on Source/GRTeclynCore/ArrayTools.hpp:
I don't really like putting this file, `GRInterval.hpp` and `Interval.hpp` under
`GRTeclynCore`. I think these small helper files should go in separate place,
maybe `CoreTools`? I disagree with @the-florist and don't think this belongs in `GridTools`/`Grid`. I wouldn't be opposed to *`Interval.hpp` being moved to `GridTools`/`Grid` though since that's the context they're used in.

mirenradia commented on Source/GRTeclynCore/Make.package:
Let's keep to the 1 file per line style.

mirenradia commented on Source/GridTools/Make.package:
Can you make this 1 file per line and in alphabetical order?

KAClough replied:
Done

mirenradia commented on Source/IOTools/MultiLevelTask.hpp:
I don't think this belongs under `IO` since you don't have to use it to do `IO`.
Having said that, it's not been ported to AMReX so we might just get rid of it later anyway.

mirenradia commented on Source/IOTools/WeylExtraction.hpp:
I don't think this belongs here. I think it might make more sense in a `GRTools`
subdirectory but I also wouldn't oppose just putting it in the `AMRInterpolator`
subdirectory.

KAClough replied:
Left as it - remove later. Added to issue.

mirenradia commented on Source/MathsTools/Make.package:
Can you make this 1 file per line and in alphabetical order?

KAClough replied:
Done

mirenradia commented on Source/MathsTools/TensorAlgebra.hpp:
It feels a bit odd putting this in a different subdirectory to `Tensor.hpp` but
I can see why you chose this.

KAClough replied:
Yes.

mirenradia reviewed (changes requested):
Thanks Katy for doing this much-needed tidy-up! It looks pretty good but, unsurprisingly, I have a few comments/suggestions (some on individual files) :upside_down_face: :

* Why did you remove the KerrBH example? Even though it's not very interesting, I guess it's potentially a simple example to look at for new users?
* What about renaming `GridTools` -> `Grid`? I think the "Tools" here is a bit redundant. I agree with @the-florist that `NullBCFill.hpp` should go here too.
* Similarly, I think `IOTools` -> `IO` makes more sense. I agree with @the-florist that we can put `AMReX_Parameters.hpp` here. We could move `SimulationParametersBase.hpp` here too or move it to ethe `CCZ4` directory since it's mostly for that.
* I don't think it makes sense to put the `Doxygen` and `Tools` directories in the `External` one. These are not "external" pieces of code like doctest is. I would be happy for `Doxygen` to move under `Tools` but it should probably live in the base directory. I'm also happy for `External` to be renamed to `external` if that satisfies your tab completion issue.

Would you also be able to clean the commit history on this branch using an interactive rebase to do the following:
* Make commit message follow our [style guide](https://github.com/GRTLCollaboration/GRChombo/wiki/Coding-conventions#github-conventions) e.g. "Fixes" -> "Fix"
* Fixup the "Fixes test ..." commit into the "Fixes name of positive chi ...." commit.
* Fixup the "runs clang format v19" commit into the "Change CCZ4 variables name and clang format" commit.

julianakwan commented on Examples/BinaryBH/GNUmakefile:
```diff
@@ -19,14 +19,14 @@ include $(AMREX_HOME)/Src/Particle/Make.package
 GRTECLYN_SOURCE = $(GRTECLYN_HOME)/Source
 
-src_dirs := $(GRTECLYN_SOURCE)/utils \
+src_dirs := $(GRTECLYN_SOURCE)/IOTools  \
```
Can these source directories be organized in alphabetical order please? Thanks!

KAClough replied:
Done

julianakwan commented on Tests/GNUmakefile:
```diff
@@ -23,13 +23,14 @@ GRTECLYN_SOURCE = $(GRTECLYN_HOME)/Source
 src_dirs := $(GRTECLYN_SOURCE)/AMRInterpolator \
             $(GRTECLYN_SOURCE)/BlackHoles \
-            $(GRTECLYN_SOURCE)/BoxUtils  \
+            $(GRTECLYN_SOURCE)/IOTools  \
```
Could you order these alphabetically as well? Thanks!

KAClough replied:
Done

julianakwan reviewed (changes requested):
Thanks very much for this PR! I think the codebase will be much tidier as a result. I just had a few comments/questions: 

-  I also found it odd that the Kerr BH example has now been removed. I imagine it wouldn't take too much effort to port but we always say this and then leave it as an open issue for ages 😅 
-  I agree that `AMReXParameters.hpp` should be moved over to `IOTools`. I briefly considered having a separate directory called `AMReXCore` or something like that where we put the bits that we need to make AMReX work like `NullBCFill.hpp` that are not strictly numerical relativity. But those were the only two files that could name! If there are more that you know of, then it might be worth doing...
- Paul was a bit unhappy that the scalar blob had been removed.

KAClough commented:
> Hi all, The PR looks quite good overall, I like the three categories you've chosen (Math, IO and Grid tools) and I think it's really clear that the Kerr example and the BH initial conditions are all in the same folder. It also leaves more room in the Examples if we want to add new matter examples in the future! I've just got some minor suggestions for sorting the files:
> 
> 1. AMReXParameters.hpp got placed in GRTeclynCore but as it's a file dealing with text input, it could also go into IOTools.

We will refactor the params in another PR so leave this for now - added to issue #1.

> 2. ArrayTools also got put in GRTeclynCore but could go into GridTools, as for example the Tensor class definition is there too and Array/Tensor functionality seems like it could be related.

Done.

> 3. Perhaps NullBCFill could also go into GridTools as the other Boundary Condition files are there.

Done.

> Of course these are just suggestions, feel free to take or leave them :)

KAClough commented:
> Thanks Katy for doing this much-needed tidy-up! It looks pretty good but, unsurprisingly, I have a few comments/suggestions (some on individual files) 🙃 :
> 
> * Why did you remove the KerrBH example? Even though it's not very interesting, I guess it's potentially a simple example to look at for new users?

I think we should keep it, but I think it will make more sense to start from the tidied up BBH example and adapt that to be a single KerrBH rather than adapt the old code. This is a good project for someone for the August hackathon - carrying the old files around until then is not very useful.

> * What about renaming `GridTools` -> `Grid`? I think the "Tools" here is a bit redundant. I agree with @the-florist that `NullBCFill.hpp` should go here too.

Done

> * Similarly, I think `IOTools` -> `IO` makes more sense. I agree with @the-florist that we can put `AMReX_Parameters.hpp` here. We could move `SimulationParametersBase.hpp` here too or move it to ethe `CCZ4` directory since it's mostly for that.

IO renamed. (for MathsTools I also removed Tools, for consistency).

We will refactor the params in another PR so I suggest we leave the Parameter reading classes where they are for now - added to issue #1.

> * I don't think it makes sense to put the `Doxygen` and `Tools` directories in the `External` one. These are not "external" pieces of code like doctest is. I would be happy for `Doxygen` to move under `Tools` but it should probably live in the base directory. I'm also happy for `External` to be renamed to `external` if that satisfies your tab completion issue.

Done

> 
> Would you also be able to clean the commit history on this branch using an interactive rebase to do the following:
> 

I suggest that when it is merged we just squash the commits, it's not very interesting to see the individual changes anyway. Then we can make a grammatically appropriate commit message :-)

KAClough commented:
> Thanks very much for this PR! I think the codebase will be much tidier as a result. I just had a few comments/questions:
> 
> * I also found it odd that the Kerr BH example has now been removed. I imagine it wouldn't take too much effort to port but we always say this and then leave it as an open issue for ages 😅

See comment to Miren above - I am sure we will find a volunteer for this in August, but it would be better to start from scratch at that stage I think.

> * I agree that `AMReXParameters.hpp` should be moved over to `IOTools`. I briefly considered having a separate directory called `AMReXCore` or something like that where we put the bits that we need to make AMReX work like `NullBCFill.hpp` that are not strictly numerical relativity. But those were the only two files that could name! If there are more that you know of, then it might be worth doing...

Yeah, I also thought about this and came to the same conclusion. For the params, since we might split them and not all of them are IO (some are more Grids) I will leave them where they are for now, but they probably mostly belong there I agree.

> * Paul was a bit unhappy that the scalar blob had been removed.

Oh :-(

KAClough committed:
Fix the things that were raised in the PR review

KAClough committed:
Merge branch 'develop' into enhancement/file_restructure

KAClough committed:
Hopefully make clang format happy

KAClough committed:
Hopefully make the annoying linter thing happy

KAClough commented:
Ok! I think it is ready to go! Can we please squash the commits into one well named one when we merge to keep Miren happy @julianakwan?

mirenradia commented:
Could you wait until I've had one final look before merging? I'm back at work tomorrow.

mirenradia reviewed (changes requested):
> I suggest that when it is merged we just squash the commits, it's not very interesting to see the individual changes anyway. Then we can make a grammatically appropriate commit message :-)

Fine, we can squash this time. You might not be so lucky next time... :stuck_out_tongue: 

I really don't like `ArrayTools.hpp` in the `Grids` subdirectory. It's got nothing to do with our grids or looping but rather it's just adding some small helper functions that should probably be in the C++ standard library. In the absence of a more well-suited place, please can you move it to the `GRTeclynCore` subdirectory?

Otherwise, I'm pretty much happy with this PR :slightly_smiling_face:

KAClough committed:
fix for miren comments

mirenradia reviewed (approved):
Looks good to me!
